src/webgpu/util/math.ts

```diff
@@ -46,9 +47,16 @@ export function diffULP(a: number, b: number): number {
   const sign_b = (u32_b & 0x80000000) !== 0;
   const masked_a = u32_a & 0x7fffffff;
   const masked_b = u32_b & 0x7fffffff;
+  const subnormal_or_zero_a = (u32_a & 0x7f800000) === 0;
+  const subnormal_or_zero_b = (u32_b & 0x7f800000) === 0;
+
+  // If the number is not subnormal, then subtract 0x7fffff to remove the
+  // subnormal region from the ULP comparison.
+  const bits_a = masked_a - (subnormal_or_zero_a ? 0 : 0x7fffff);
```

Comment:
This seems wrong.  

From [wikipedia](https://en.wikipedia.org/wiki/Unit_in_the_last_place)

> Another definition, suggested by John Harrison, is slightly different: ulp(x)  is the distance between the two closest straddling floating-point numbers _a_ and _b_ (i.e., those with a &leq; x &leq; b  and  a &ne; b,  assuming that the exponent range is not upper-bounded.[3][4] 

So, for example when x = 0, then `a` and `b` can be selected as `a` = x = 0, and `b` is the smallest subnormal number. We may as well pick `b` having the same sign as the zero.

This avoids the divide-by-zero problem near zero.

Then diffULP(0,y), assuming y > 0, would be well defined.   If y is finite (either normal or subnormal), then the answer is going to be:
* 0 when y is zero
* masked_y  as if computed around line 48 and 49, when y != 0

As luck has it, both cases are the `masked_y` value.

The original issue #837 says negative zero was the problem.

WGSL says:

> Implementations may ignore the sign of a zero. That is, a zero with a positive sign may behave like a zero a with a negative sign, and vice versa.

Can we accommodate this in diffULP by patching up the sign when the magnitude is zero:

```js
if ( masked_a === 0 && masked_b === 0) { return 0; } 
if (masked_a === 0) { sign_a = sign_b; }
if (masked_b === 0) { sign_b = sign_a; }
```

Reply:
As discussed internally, I think we make the fix with just replacing these two lines:

```ts
const bits_a = masked_a - (subnormal_or_zero_a ? 0 : 0x7fffff);
const bits_b = masked_b - (subnormal_or_zero_b ? 0 : 0x7fffff);
```

with:

```ts
const bits_a = subnormal_or_zero_a ? 0 : (masked_a - 0x800000);
const bits_b = subnormal_or_zero_b ? 0 : (masked_b - 0x800000);
```

David would like more comments on this function to describe the subnormal behavior (it's just round-to-zero), and some more tests around these cases.

Reply:
Rationale:  The John Harrison definition of ULP(X) requires us to find the two floating point values a and b that straddle X.  When an implementation is allowed to flush subnormals to zero, then in then   {a,b} = {0, smallest_normal_number} (as sets).  And so ULP(0) and ULP(any subnormal)  will be equal to the smallest normal floating point number.

Reply:
Expanding this out:

Then diffULP(x,y) is "the number of ULP(min(|x|,|y|)) steps between x and y"

So diffULP(0,subnormal) = 0  (I think).  and diffULP(0,smallest_normal) = 1

Reply:
I have updated the code and extended the tests. The change for `bits_*` that I actually used is this:

```
const bits_a = subnormal_or_zero_a ? 0 : (masked_a - 0x7fffff);
const bits_b = subnormal_or_zero_b ? 0 : (masked_b - 0x7fffff);
```

since using `0x800000` would push the smallest normal to 0, diffULP(0, smallest normal) was returning instead of 1.